### PR TITLE
Phase 5: Full Transolver-3 — Complete Model from Latest Paper (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Build the **complete Transolver-3** model from the latest paper (arXiv 2602.04940, Feb 2026), replacing the ENTIRE current Transolver. This is per human team direction to try full model replacements of the latest architectures.

Key Transolver-3 innovations over our current model:
1. **Faster slice/deslice** — reorder matrix multiplications to reduce memory
2. **Per-point adaptive temperature** (from Transolver++) — each node gets its own slice routing temperature
3. **Gumbel-Softmax slice assignment** — discrete-like routing during training
4. **Amortized training on mesh subsets** — train on random subsets of nodes per step

**This is a FULL replacement**, not modifications on top of the existing model. Delete the current Transolver classes and implement fresh from the paper.

## Instructions

Delete `Physics_Attention_Irregular_Mesh`, `TransolverBlock`, and `Transolver` classes. Implement the Transolver-3 architecture fresh:

1. **New slice attention** with per-point adaptive temperature and Gumbel-Softmax
2. **Memory-efficient slice/deslice** — compute `slice_token = (X^T @ W) @ V` instead of materializing full `W` matrix
3. **Node subsampling** — randomly sample 50-75% of nodes per training step (always keep surface nodes)
4. **Pressure-first head** — keep this from our proven improvement

### GPU Sweep:
| GPU | subsample | n_hidden | gumbel_scale | Notes |
|-----|-----------|----------|-------------|-------|
| 0 | 0.75 | 192 | 0.1 | Default T3 |
| 1 | 0.50 | 192 | 0.1 | More aggressive subsampling |
| 2 | 0.75 | 256 | 0.1 | Wider (memory savings allow) |
| 3 | 0.75 | 192 | 0.05 | Gentler Gumbel |
| 4 | 0.75 | 192 | 0.2 | Stronger Gumbel |
| 5 | 0.50 | 256 | 0.1 | Combined |
| 6 | 0.75 | 192 | 0.1 | seed 43 |
| 7 | 1.00 | 192 | 0.0 | Baseline (no T3 changes) |

## Baseline
val/loss 0.401, p_in 12.95, p_oodc 8.40, p_tan 33.8, p_re 24.7

### Reproduce:
```bash
python train.py --agent edward --wandb_name "edward/baseline" \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --cosine_T_max 180 --disable_pcgrad \
  --pressure_first --pressure_deep
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Round 2 Results (--compile_mode none)

### Summary

With torch.compile disabled, all experiments now start training immediately (no 114-min compile overhead). However, **without compile, per-epoch throughput drops ~2x** (115-161s vs 50-75s compiled), so both T3 and baseline reach fewer total epochs. The T3 models reached 50 epochs and baseline 78 — a fairer comparison, but neither matches the compiled R1 baseline quality (168 epochs, p_in=12.6).

**T3 is still worse than the uncompiled baseline** at comparable epoch counts, and the uncompiled baseline itself is much worse than the compiled baseline.

### R2 Results Table

| GPU | Config | Run ID | Epochs | VRAM | val_in | val_tan | val_oodc | val_re |
|-----|--------|--------|--------|------|--------|---------|----------|--------|
| 0 | T3 default (sub=0.75, h=192, g=0.1) | 1wm9dzul | 50 | 38.0 GB | 0.838 | 2.345 | 0.919 | 0.858 |
| 1 | T3 sub=0.50 | devrcciu | 50 | 25.6 GB | 1.308 | 5.490 | 0.723 | 0.674 |
| 2 | T3 wide=256 | tk5dla74 | 50 | 46.7 GB | 0.861 | 1.933 | 0.678 | 0.621 |
| 3 | T3 gumbel=0.05 | npfqymby | 50 | 38.0 GB | 0.862 | 2.183 | 0.631 | 0.563 |
| 4 | T3 gumbel=0.2 | qo8z7g4a | 50 | 38.0 GB | 0.884 | 2.094 | 0.698 | 0.654 |
| 5 | T3 wide+sub (256, 0.50) | 2ht6fu2y | 50 | 31.4 GB | 0.632 | 2.319 | 0.631 | 0.558 |
| 6 | T3 seed 43 | cx3aoo6e | 50 | 38.0 GB | 0.915 | 2.523 | 0.980 | 0.916 |
| 7 | **Baseline (no compile)** | zweku0mp | **78** | 45.7 GB | **0.437** | **1.200** | **0.307** | **0.263** |

**Baseline surface MAE (no compile, 78 epochs):** p_in=21.9, p_tan=41.2, p_oodc=16.1, p_re=29.5
**R1 compiled baseline (168 epochs):** p_in=12.6, p_tan=33.2, p_oodc=8.2, p_re=24.5

### Key Findings

1. **Disabling compile helps T3 get training time** but hurts BOTH T3 and baseline equally. Without compile, everything is ~2x slower per epoch (115-161s vs 50-75s compiled). The uncompiled baseline (p_in=21.9) is 74% worse than the compiled one (12.6).

2. **T3 still worse than uncompiled baseline**: At 50 epochs (T3) vs 78 (baseline), the best T3 variant (GPU 3, gumbel=0.05: val_oodc=0.631) is still 2x worse than the uncompiled baseline (val_oodc=0.307). The baseline converges faster per epoch.

3. **Subsampling too aggressive**: GPU 1 (50% subsample) was worst again — dropping half the nodes hurts too much even though it speeds up training.

4. **The core issue**: The T3 memory-efficient slice/deslice uses `x_expanded = x.unsqueeze(1).expand(-1, H, -1, -1)` which creates a [B, H, N, C] tensor at full feature dim C instead of dim_head D. This is actually LARGER than the original approach, negating the memory savings. The efficiency gain only materializes when C < D (true for very wide models, not for ours where C ≈ D).

### Conclusion

The T3 architecture changes (per-point temperature, Gumbel-Softmax, memory-efficient slice/deslice) **do not improve over the baseline** at our problem scale (~50K nodes, 192-dim):
- Per-point temperature adds parameters and compile complexity without clear accuracy benefit
- Gumbel-Softmax noise provides no measurable improvement over learned softmax temperature
- The "memory-efficient" slice rewrite is not actually more efficient at our feature dimensions
- torch.compile is critical for throughput (2x speedup), but the new code patterns are hostile to torch.compile's tracer

### Commands Used

```bash
BASE="--agent edward --field_decoder --adaln_output --use_lion --lr 2e-4 --cosine_T_max 180 --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 --disable_pcgrad --pressure_first --pressure_deep --val_every 2 --seed 42 --compile_mode none"

# GPU 0-6: T3 variants with --t3_mode --compile_mode none
# GPU 7: Baseline with --compile_mode none
```

W&B group: `phase5-t3-r2` in project `wandb-applied-ai-team/senpai-v1`
